src/app/clusters/actions-server/actions-server.cpp

```diff
@@ -0,0 +1,418 @@
+        HandleCommand<Commands::ResumeAction::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & commandData) { HandleResumeAction(ctx, commandData); });
+        return;
+    case Actions::Commands::EnableAction::Id:
+        HandleCommand<Commands::EnableAction::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & commandData) { HandleEnableAction(ctx, commandData); });
+        return;
+    case Actions::Commands::EnableActionWithDuration::Id:
+        HandleCommand<Commands::EnableActionWithDuration::DecodableType>(
+            handlerContext,
+            [this](HandlerContext & ctx, const auto & commandData) { HandleEnableActionWithDuration(ctx, commandData); });
+        return;
+    case Actions::Commands::DisableAction::Id:
+        HandleCommand<Commands::DisableAction::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & commandData) { HandleDisableAction(ctx, commandData); });
+        return;
+    case Actions::Commands::DisableActionWithDuration::Id:
+        HandleCommand<Commands::DisableActionWithDuration::DecodableType>(
+            handlerContext,
+            [this](HandlerContext & ctx, const auto & commandData) { HandleDisableActionWithDuration(ctx, commandData); });
+        return;
+    }
+}
+
+CHIP_ERROR ActionsServer::SetActionList(EndpointId aEndpoint, const ActionStructStorage & aAction)
+{
+    Delegate * delegate = GetDelegate(aEndpoint);
+    if (delegate == nullptr)
+    {
+        ChipLogError(Zcl, "Actions delegate is null!");
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+
+    // Read through the list to find and update the existing action
+    for (uint16_t i = 0; i < kMaxActionListLength; i++)
+    {
+        ActionStructStorage existingAction;
+        ReturnErrorOnFailure(delegate->ReadActionAtIndex(i, existingAction));
+
+        if (existingAction.actionID == aAction.actionID)
+        {
+            existingAction.Set(aAction.actionID, aAction.name, aAction.type, aAction.endpointListID, aAction.supportedCommands,
+                               aAction.state);
+
+            delegate->OnActionListChanged(aEndpoint, aAction);
+            mMatterContext.MarkDirty(aEndpoint, Attributes::ActionList::Id);
+            return CHIP_NO_ERROR;
+        }
+    }
+
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+CHIP_ERROR ActionsServer::SetEndpointList(EndpointId aEndpoint, const EndpointListStorage & aEpList)
+{
+    Delegate * delegate = GetDelegate(aEndpoint);
+    if (delegate == nullptr)
+    {
+        ChipLogError(Zcl, "Actions delegate is null!");
+        return CHIP_ERROR_INCORRECT_STATE;
+    }
+
+    // Read through the list to find and update the existing endpoint list
+    for (uint16_t i = 0; i < kMaxEndpointListLength; i++)
+    {
+        EndpointListStorage existingEpList;
+        ReturnErrorOnFailure(delegate->ReadEndpointListAtIndex(i, existingEpList));
```

Comment:
What if there are fewer than kMaxEndpointListLength lists?

Reply:
Updated the behaviour to break and return proper error.